presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,703 @@
+
+            initializeTempStorage();
+            context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+            ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("dispatcher-query-%s"));
+            closer.register(coreExecutor::shutdownNow);
+            executor = listeningDecorator(coreExecutor);
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == OutputBuffers.BufferType.SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        requireNonNull(bufferId, "outputBufferId is null");
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+
+        acknowledge(bufferId, token);
+        long currentSequenceId;
+
+        try {
+            synchronized (this) {
+                currentSequenceId = sequenceId.get();
+
+                // process the request if we have no more data coming in, have data to read, or if this is an outdated request
+                if (noMorePages.get() || !handleInfoQueue.isEmpty() || !pages.isEmpty() || currentSequenceId != token) {
+                    return immediateFuture(processRead(token, maxSize));
```

Comment:
As discussed, we'll want to make this lazy and perform the work with the help of an executor.

Reply:
Done